exporter/src/main/scala/uk/gov/nationalarchives/consignmentexport/BagMetadata.scala

```diff
@@ -0,0 +1,102 @@
+package uk.gov.nationalarchives.consignmentexport
+
+import java.util.UUID
+
+import cats.effect.IO
+import gov.loc.repository.bagit.domain.Metadata
+import graphql.codegen.GetConsignmentExport.getConsignmentForExport.GetConsignment
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import org.keycloak.representations.idm.UserRepresentation
+import uk.gov.nationalarchives.consignmentexport.BagMetadata._
+import uk.gov.nationalarchives.consignmentexport.Config.Configuration
+
+class BagMetadata(graphQlApi: GraphQlApi, keycloakClient: KeycloakClient)(implicit val logger: SelfAwareStructuredLogger[IO]) {
+  def getBagMetadata(consignmentId: UUID, config: Configuration): IO[Metadata] = for {
+   consignment <- graphQlApi.getConsignmentMetadata(config, consignmentId)
+   consignmentDetails = consignment match {
+     case Some(consignment) => getConsignmentDetails(consignment)
+     case None => throw new RuntimeException(s"No consignment metadata found for consignment $consignmentId")
+   }
+
+  } yield generateMetadata(consignmentDetails)
+
+  private def getConsignmentDetails(consignment: GetConsignment): Map[String, Option[String]] = {
+    val seriesCode = for {
+      series <- consignment.series
+      sc <- series.code
+    } yield sc
+
+    val bodyCode = for {
+      body <- consignment.transferringBody
+      bc <- body.code
+    } yield bc
+
+    val startDatetime = for {
+      createdDate <- consignment.createdDatetime
+      cd = createdDate.toString
+    } yield cd
+
+    val completedDatetime = for {
+      completedDate <- consignment.transferInitiatedDatetime
+      cd = completedDate.toString
+    } yield cd
+
+    val exportDatetime = for {
+      exportDate <- consignment.exportDatetime
+      ed = exportDate.toString
+    } yield ed
+
+    val contactName = getContactName(consignment.userid)
+
+    Map(
+      ConsignmentSeriesKey -> seriesCode,
+      SourceOrganisationKey -> bodyCode,
+      ConsignmentStartDateKey -> startDatetime,
+      ConsignmentCompletedDateKey -> completedDatetime,
+      ConsignmentExportDateKey -> exportDatetime,
+      ContactNameKey -> Some(contactName)
+    )
+  }
+
+  private def generateMetadata(details: Map[String, Option[String]]): Metadata = {
+    val metadata = new Metadata
+
+    details.map(e => {
+      e._2 match {
+        case Some(_) => metadata.add(e._1, e._2.get)
+        case None => //For now do nothing is property is missing
```

Comment:
When we discussed this the other day, I was thinking that we were fetching all of the metadata from the API as a big list, so we wouldn't know what the expected values were. Which I realise now is silly, because of course things like the creation date are not stored in the metadata table.

If we're fetching things by name, then we might need to think more carefully about this decision.

Maybe we should discuss this with Sam tomorrow, since he's looking into how to structure the API response for the file metadata?

Reply:
Given the discussion we had with Sam, is it worth retaining this additional step of putting the consignment metadata into a map, then creating the Bagit Metadata object from this map?